cmd/cmrel/cmd/validate_gomod.go

```diff
@@ -0,0 +1,164 @@
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+
+	"github.com/spf13/cobra"
+	flag "github.com/spf13/pflag"
+	"golang.org/x/mod/modfile"
+)
+
+const (
+	validateGoModCommand         = "validate-gomod"
+	validateGoModDescription     = "Parse cert-manager go.mod files to ensure they're importable"
+	validateGoModLongDescription = `Parses cert-manager go.mod files to enforce certain requirements on them.
+
+NB: We talk about "core" referring to the main go.mod file for the cert-manager
+packge, and "sub" referring to go.mod files for modules we don't expect people
+to import.
+
+Ensures that:
+
+- Any replace directives for 3rd party dependencies in core are also present
+  used for all subs which also have those dependencies, preventing drift
+- All subs use an invalid version of cert-manager in their go.mod so they're
+  forced to rely on replace directives pointing to the local module in the repo
+- All modules declare the same version of Golang`
+)
+
+var (
+	validateGoModExample = fmt.Sprintf(`To validate a local checkout:
+
+%s %s --path <path-to-checkout>`, rootCommand, validateGoModCommand)
+)
+
+type validateGoModOptions struct {
+	// Path is the path of the cert-manager checkout to validate
+	Path string
+}
+
+func (o *validateGoModOptions) AddFlags(fs *flag.FlagSet, markRequired func(string)) {
+	fs.StringVar(&o.Path, "path", "", "Path of cert-manager checkout to validate")
+
+	markRequired("path")
+}
+
+func (o *validateGoModOptions) print() {
+	log.Printf("%s options:", validateGoModCommand)
+	log.Printf("  Path: %q", o.Path)
+}
+
+func validateGoModCmd(rootOpts *rootOptions) *cobra.Command {
+	o := &validateGoModOptions{}
+
+	cmd := &cobra.Command{
+		Use:          validateGoModCommand,
+		Short:        validateGoModDescription,
+		Long:         validateGoModLongDescription,
+		Example:      validateGoModExample,
+		SilenceUsage: true,
+		PreRun: func(cmd *cobra.Command, args []string) {
+			o.print()
+			log.Printf("---")
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return runValidateGoMod(rootOpts, o)
+		},
+	}
+
+	o.AddFlags(cmd.Flags(), mustMarkRequired(cmd.MarkFlagRequired))
+
+	return cmd
+}
+
+func runValidateGoMod(rootOpts *rootOptions, o *validateGoModOptions) error {
+	rootGoModFile := filepath.Join(o.Path, "go.mod")
+	rootGoModData, err := os.ReadFile(rootGoModFile)
+	if err != nil {
+		return fmt.Errorf("failed to read root go.mod file: %s", err.Error())
+	}
+
+	rootModule, err := modfile.Parse(rootGoModFile, rootGoModData, nil)
+	if err != nil {
+		return fmt.Errorf("failed to parse root go.mod file: %s", err.Error())
+	}
+
+	// coreImportName := rootModule.Module.Mod
+	coreGoVersion := rootModule.Go.Version
+
+	submodules := []string{
+		"cmd/acmesolver/go.mod",
+		"cmd/cainjector/go.mod",
+		"cmd/ctl/go.mod",
+		"cmd/controller/go.mod",
+		"cmd/webhook/go.mod",
+		"test/integration/go.mod",
+		"test/e2e/go.mod",
```

Comment:
can we do a file-system search here instead? That would make it harder to fail (because of an outdated list eg.)

Reply:
Of course! This is very much still WIP and is missing most of the actually important scanning functionality. A hardcoded list is quicker to write so I get to work on that core logic faster!